p2p/enr/enr.go

```diff
@@ -0,0 +1,396 @@
+	if err != nil {
+		return err
+	}
+
+	seq, err := s.Uint()
+	if err != nil {
+		return err
+	}
+
+	var records []record
+
+	// read key/value pairs until we reach rlp.EOL
+	for _, _, err = s.Kind(); err == nil; _, _, err = s.Kind() {
+		key, err2 := s.Bytes()
+		if err2 != nil {
+			return err2
+		}
+
+		value, err2 := s.Bytes()
+		if err2 != nil {
+			return err2
+		}
+
+		records = append(records, record{k: key, v: value})
+	}
+
+	if err != rlp.EOL {
+		return err
+	}
+
+	e.signature = signature
+	e.raw = data
+	e.seq = uint32(seq)
+	e.records = records
+
+	err = e.verifySignature()
+	if err != nil {
+		return err
+	}
+
+	e.dirty = false
+
+	return nil
+}
+
+func (e *ENR) Sign(privkey *ecdsa.PrivateKey) error {
+	e.seq = e.seq + 1
+
+	e.SetID(ID_SECP256k1_KECCAK)
+
+	pk := (*btcec.PublicKey)(&privkey.PublicKey).SerializeCompressed()
+	e.SetSecp256k1(pk)
+
+	var err error
+	e.signature, e.raw, err = e.SignAndEncode(privkey)
+	if err != nil {
+		return err
+	}
+
+	// mark record ready for encoding
+	e.dirty = false
+
+	return nil
+}
+
+func (e *ENR) SignAndEncode(privkey *ecdsa.PrivateKey) ([]byte, []byte, error) {
+	content, err := e.SerialisedContent()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	digest := crypto.Keccak256Hash(content)
+
+	signature, err := crypto.Sign(digest.Bytes(), privkey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	blob, err := rlp.EncodeToBytes(signature)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	raw := append(blob, content...)
+
+	return signature, raw, nil
+}
+
+func (e *ENR) SerialisedContent() ([]byte, error) {
+	var buffer bytes.Buffer
+
+	blob, err := rlp.EncodeToBytes(e.seq)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = buffer.Write(blob)
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Slice(e.records, func(i, j int) bool {
+		return bytes.Compare(e.records[i].k, e.records[j].k) < 0
+	})
+
+	for _, r := range e.records {
+		kk, err := rlp.EncodeToBytes(r.k)
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = buffer.Write(kk)
+		if err != nil {
+			return nil, err
+		}
+
+		vv, err := rlp.EncodeToBytes(r.v)
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = buffer.Write(vv)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return wrapList(buffer.Bytes()), nil
+}
+
+func (e *ENR) verifySignature() error {
+	id, err := e.GetID()
+	if err != nil {
+		return err
+	}
+
+	// currently "secp256k1-keccak" is the only known identity scheme
+	if id != ID_SECP256k1_KECCAK {
+		return errors.New("unknown identity scheme")
+	}
+
+	// get publickey from record
+	blob, err := e.GetSecp256k1()
+	if err != nil {
+		return err
+	}
+
+	pk, err := btcec.ParsePubKey(blob, btcec.S256())
+	if err != nil {
+		return err
+	}
+	pubkey1 := pk.SerializeUncompressed()
+
+	// get publickey from message and signature
+	content, err := e.SerialisedContent()
+	if err != nil {
+		return err
+	}
+
+	digest := crypto.Keccak256Hash(content)
+	pubkey2, err := crypto.Ecrecover(digest.Bytes(), e.signature)
```

Comment:
You can use `btcec.Signature.Verify` instead: https://godoc.org/github.com/btcsuite/btcd/btcec#example-package--VerifySignature